test/e2e/mongodbtests/mongodbtests.go

```diff
@@ -276,6 +290,49 @@ func Connect(mdb *mdbv1.MongoDB, opts *options.ClientOptions) error {
 	})
 }
 
+func ConnectAndVerifyExpectedCustomRoles(mdb *mdbv1.MongoDB, database, username, password string, expectedRoles []automationconfig.CustomRole) func(t *testing.T) {
```

Comment:
it looks like a lot of this functionality we actually already have! If you look at the `tester, err := FromResource(t, mdb)` call, that function call configures the underlying golang client with the correct SCRAM credentials for that resource, and there is a method [hasAdminParameter](https://github.com/mongodb/mongodb-kubernetes-operator/blob/5fa5815cd5e458121c4c1e99361fb45cea5c703d/test/e2e/util/mongotester/mongotester.go#L138) which can be used to check for a specific field in the `admin` db. 

The end result would look something like 

```golang
t.Run("Custom Role was created ", tester.VerifyRoles(&mdb, roles))
```

since the client configuration should already be done.

Reply:
The functionality is similar; however, there are a few key differences. The main one is that I am running a different command than `getAdminSetting` in order to filter out builtin roles:
```golang
RunCommand(ctx, bson.D{
  {"rolesInfo", 1},            // nolint
  {"showPrivileges", true},    // nolint
  {"showBuiltinRoles", false}, // nolint
}).Decode(&result)
```
instead of
```golang
RunCommand(context.TODO(), bson.D{{Key: "getParameter", Value: 1}, {Key: key, Value: 1}})
```
Another difference is that because it is not possible to predict the order in which roles will be returned from mongo, and because it may not match the expected order, I am now using `assert.ElementsMatch(t, result.Roles, expectedRoles)` to compare the array result, whereas `hasAdminSetting` uses `reflect.DeepEqual(expectedValue, actualValue)`. Note that I added additional roles to the e2e test to ensure that all valid formats of resources are accepted.

I also prefer using the strong typing of decoding the result into a struct instead of a map using:
```golang
// CustomRolesResult is a type to decode the result of getting rolesInfo.
type CustomRolesResult struct {
	Roles []automationconfig.CustomRole
}
```

That being said, I will spend some time to see if I can refactor the tester nicely to reduce code duplication while keeping at least the first two requirements for this function intact.

Reply:
This is done! I refactored to reuse the old code. I think it looks much cleaner now, so thank you for pointing this out!